// ts-osc/.eslintrc.js
module.exports = {
  parser: "@typescript-eslint/parser", // Specifies the ESLint parser

  parserOptions: {
    ecmaVersion: 2020, // Allows for the parsing of modern ECMAScript features
    sourceType: "module" // Allows for the use of imports
  },

  extends: [
    "plugin:@typescript-eslint/recommended",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended"
  ],
  rules: {
    "@typescript-eslint/no-shadow": ["error", { "allow": ["err"] }],
    "no-console": "off",
    "no-use-before-define": "off",
    "no-constant-condition": "off",
    "import/no-unresolved": "off",
    "import/extensions": "off",
    "class-methods-use-this": "off",
    "@typescript-eslint/array-type": [2, { "default": "array" }],
    "@typescript-eslint/consistent-type-definitions": [2, "interface"],
    "@typescript-eslint/no-explicit-any": "error",
    "@typescript-eslint/no-for-in-array": "error",
    "@typescript-eslint/no-inferrable-types": "error",
    "@typescript-eslint/no-use-before-define": "warn",
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": "warn",
    "no-underscore-dangle": "off",
    "prefer-destructuring": "off"
  }
};
